### DISCUSSION TITLE
Issue with Storing Multiple Bonding Information in NVS

### DISCUSSION THREAD

**Discussion by bblodget** [h2zero/NimBLE-Arduino#611] [Q&A] [answered]
I am encountering a problem with storing multiple bonding information in NVS using the NimBLE-Arduino library. Despite the expectation that the ESP32 should manage multiple bonds, it appears to only retain bonding information for the last paired device, overwriting previous entries.  The device I'm using is a esp32-pico-d4.

### Setup

- **NimBLE-Arduino Library Version:** Initially on the head of the release/1.4 branch, but also tested on tags 1.4.1, 1.4.0, and 1.3.0
- **nimconfig.h config**: Default config with MAX_CONNECTIONS=3, MAX_BONDS=3, MAX_CCCDS=8, NVS_PERSIST=1
- BLEDevice::setSecurityAuth(true, true, true);

### Issue Description

The behavior observed is as follows:

1. **First Pairing:** Successfully pair with a Windows device. Bonding information is stored in NVS. NimBLEDevice::getNumBonds() returns 1.  Able to connect to Windows device.
2. **Second Pairing:** Pair with a Mac device. The previously stored bonding information (Windows device) seems to be overwritten. NimBLEDevice::getNumBonds() returns 1.  Able to connect to the Mac device.  The Windows device will attempt to connect and then disconnect.

Any help or guidance would be greatly appreciated.  Thank you!

**Comment by h2zero** [comment]
I believe there is a bug in the core code that may be causing this. I'll investigate and get back to you.

**Comment by h2zero** [comment]
I have done a quick test and cannot reproduce but I'd like to ask you to try erasing the flash completely and then trying again and let me know what the result is.

**Comment by bblodget** [comment]
So I erased the flash and had the same results.  I've made a smaller test program to try and capture the problem.  It is interesting using the test program I was able to connect two devices, and two device had been bonded.  So that is progress.  However when I try and bond the third device, it seems to momentary succeed, and I see three bonded devices.  But then the first device disconnect for no apparent reason and then it says there are only two bonded devices.  The first device then continues to try to reconnect but fails since the bonding info is missing.  This seems to be similar to the problem I had before, but I was limited to just one bonded device, but now I seem to be able to get two bonded devices. 

I've created a gist for my test case, [nimble_test.cpp](https://gist.github.com/bblodget/5b161a3a8f6da4d98f5e52ee5ba0a3ce).

I've also created a gist for my output log, [com_log5.txt](https://gist.github.com/bblodget/ac2c3e3a071caef0ef894ffcd1075c24).

I erased the flash before running this test.

**Comment by h2zero** [comment, marked as answer]
Thank you for getting back to me, I suspect the issue is due to not enough CCCD storage. Please try increasing `MAX_CCCDS` from 8 to 9 and erasing the flash to test again.

**Reply by bblodget** [reply]
Yes!  Increasing the MAX_CCCDS solved the problem.  In my test code I found I had to increase it to 10.  In my full app I had to increase it to 13, to store the bonding info for three devices.  Thank you!

**Reply by h2zero** [reply]
Awesome, glad that was the fix 🙂
